src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -1011,6 +1017,14 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
                 let ty = ty.subst(tcx, substs);
                 let ty = self.normalize(ty, locations);
 
+                let mut projected_ty = PlaceTy::from_ty(ty);
+                for proj in &user_ty.projs {
+                    projected_ty = projected_ty.projection_ty(
+                        tcx, proj, |this, field, &()| this.field_ty(tcx, field));
```

Comment:
We will need to normalize here, I think

Reply:
Okay I think this will serve as a basis for a test showing that we need to normalize as we project the fields:

```rust
struct Single<T> { value: T }

trait Indirect {
    type U;
}

impl<'a> Indirect for &'a u32 {
    type U = &'a u32;
}

fn static_to_a_to_static_through_struct_2<'a>(_x: &'a u32) -> &'static u32 {
    let Single { value: y }: Single<<&'a u32 as Indirect>::U> = Single { value: &22 };
    y //~ ERROR
}
```
```

Reply:
(well, either that, or the example exposes bugs elsewhere. Ha ha ... ha. ha.)

Reply:
right, the previous example looks an awful like an embedding of #54940. So maybe thats not going to get me anywhere.

Reply:
(If it comes down to it, I'll happily put in the code to normalize in the callback itself. But I would like to spend a bit longer trying to identify a test case that the current PR breaks on...)

Reply:
ah this one looks promising:

```rust
trait Indirect { type Assoc; }
struct StaticU32;
impl Indirect for StaticU32 { type Assoc = &'static u32; }

struct Single2<T: Indirect> { value: <T as Indirect>::Assoc }

fn static_no_initializer_2() {
    let x = 22;
    let Single2 { value: mut _y }: Single2<StaticU32>;
    _y = &x; //~ ERROR
}
```